File: kafka-streams-java/4-user-event-enricher/src/main/java/com/github/simplesteph/udemy/kafka/streams/GameGenerator.java
package com.github.simplesteph.udemy.kafka.streams;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.simplesteph.udemy.kafka.streams.internal.*;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.Producer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.Serdes;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;

public class GameGenerator extends Generator {

    private static ObjectMapper mapper = new ObjectMapper();

    private static Logger logger = LoggerFactory.getLogger(GameGenerator.class);

    public static void main(String[] args) throws InterruptedException {
        Properties properties = new Properties();

        properties.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "127.0.0.1:9092");
        properties.setProperty(ProducerConfig.ACKS_CONFIG, "all");

        properties.setProperty(ProducerConfig.RETRIES_CONFIG, "3");
        properties.setProperty(ProducerConfig.LINGER_MS_CONFIG, "1");
        properties.setProperty(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, "true");

        Producer<String, String> producer =
                new KafkaProducer<>(properties, Serdes.String().serializer(), Serdes.String().serializer());

        /* add a user to this list to see him in the list of purchases */
        ArrayList<User> users = new ArrayList<>( Arrays.asList(
                new User("jdoe", "john", "Doe", "john.doe@gmail.com"),
                new User("jojo", "Johnny", "Doe", "johnny.doe@gmail.com"),
                new User("simplesteph", "Stephane", "Maarek", null),
                new User("alice", "Alice", null, null),
                new User("will00", "Will", "Byers", "will@st.com"),
                new User("kali01", "Kali", "Eight", "kali@st.com"),
                new User("lucas03", "Lucas", "Sinclair", "lucas@st.com"),
                new User("jon04", "Jonathan", "Byers", "jonathan@st.com"),
                new User("dustin05", "Dustin", "Henderson", "dustin@st.com"),
                new User("mike06", "Mike", "Wheeler", "mike@st.com")
        ));

        users
                .stream()
                .map(GameGenerator::userRecord)
                .forEach(producer::send);

        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            producer.flush();
            producer.close();
        }));

        producer.flush();

        int i = 0;
        while (true){
            Thread.sleep(1000L);

            if (i % 10 == 0) logger.info(String.format("Generating the %sth Purchase", i));

            // produce purchase
            Optional<ProducerRecord<String, String>> maybeRecord =

                    Optional.ofNullable(nextUser(users))
                            .map((user) -> {

                                Purchase value = nextPurchase(user);

                                return purchaseRecord(value);
                            }
                    );

            if (i % 10 == 0) logger.info(String.format("Sending the %sth Producer Record", i));

            maybeRecord.ifPresent(producer::send);

            i += 1;
        }
    }

    private static ProducerRecord<String, String> userRecord(User user){
        return new ProducerRecord<>("user-table", user.getLogin(), writeJson(user));
    }

    private static ProducerRecord<String, String> userDeletionRecord(String login){
        return new ProducerRecord<>("user-table", login, null);
    }

    private static ProducerRecord<String, String> purchaseRecord(Purchase purchase){
        return new ProducerRecord<>("user-purchases", purchase.getUser(), writeJson(purchase));
    }

    private static String writeJson(Object value) {
        try {
            return mapper.writeValueAsString(value);
        } catch (JsonProcessingException e) {
            e.printStackTrace();
            return "{}";
        }
    }

}

File: kafka-streams-scala/6-avro-messages-scala/src/main/scala/com/github/simplesteph/udemy/kafka/streams/GameGenerator.scala
package com.github.simplesteph.udemy.kafka.streams

import java.util.Properties

import com.github.simplesteph._
import com.github.simplesteph.udemy.kafka.streams.internal.Generator
import io.confluent.kafka.serializers.AbstractKafkaAvroSerDeConfig
import io.confluent.kafka.streams.serdes.avro.SpecificAvroSerde
import org.apache.kafka.clients.producer.{KafkaProducer, ProducerConfig, ProducerRecord}
import org.slf4j.{Logger, LoggerFactory}

import scala.collection.JavaConverters._

object GameGenerator extends App with Generator {

  private val logger: Logger = LoggerFactory.getLogger(getClass)

  val properties = new Properties
  // kafka bootstrap server
  properties.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "127.0.0.1:9092")
  // producer acks
  properties.setProperty(ProducerConfig.ACKS_CONFIG, "all") // strongest producing guarantee

  properties.setProperty(ProducerConfig.RETRIES_CONFIG, "3")
  properties.setProperty(ProducerConfig.LINGER_MS_CONFIG, "1")
  // leverage idempotent producer from Kafka 0.11 !
  properties.setProperty(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, "true") // ensure we don't push duplicates

  val userSerde = new SpecificAvroSerde[User]()
  val userKeySerde = new SpecificAvroSerde[UserKey]()
  val purchaseSerde = new SpecificAvroSerde[Purchase]()
  val purchaseKeySerde = new SpecificAvroSerde[PurchaseKey]()

  userKeySerde :: purchaseKeySerde :: Nil foreach {
    _.configure(Map(AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG -> "http://127.0.0.1:8081").asJava, true)
  }
  userSerde :: purchaseSerde :: Nil foreach {
    _.configure(Map(AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG -> "http://127.0.0.1:8081").asJava, false)
  }

  val userProducer = new KafkaProducer[UserKey, User](
    properties,
    userKeySerde.serializer(),
    userSerde.serializer()
  )

  val purchaseProducer = new KafkaProducer[PurchaseKey, Purchase](
    properties,
    purchaseKeySerde.serializer(),
    purchaseSerde.serializer()
  )

  val users = Vector[User](
    User("jdoe", "john", Some("Doe"), Some("john.doe@gmail.com")),
    User("jojo", "Johnny", Some("Doe"), Some("johnny.doe@gmail.com")),
    User("simplesteph", "Stephane", Some("Maarek")),
    User("alice", "Alice"),
    User("will01", "Will", Some("Byers"), Some("will@st.com")),
    User("kali02", "Kali", Some("Eight"), Some("kali@st.com")),
    User("lucas03", "Lucas", Some("Sinclair"), Some("lucas@st.com")),
    User("jon04", "Jonathan", Some("Byers"), Some("jonathan@st.com")),
    User("dustin05", "Dustin", Some("Henderson"), Some("dustin@st.com")),
    User("mike06", "Mike", Some("Wheeler"), Some("mike@st.com"))
  )

  /* produce all users in a compacted topic (to build a ktable)  */
  users
    .map(user => new ProducerRecord[UserKey, User]("avro-user-table", UserKey(user.login), user))
    .foreach(userProducer.send)

  userProducer.flush()

  var i = 0

  /* produce purchase in topic (to build a kstream) */
  while (true) {

    Thread.sleep(1000)

    if (i % 10 == 0) logger info s"Generating the ${i}th Purchase"

    for {
      game <- nextGame
      user <- nexUser(users)
      id <- nextPurchaseId.orElse(Some("X"))
      isTwoPlayer <- nextIsTwoPlayer.orElse(Some(true))
    } yield {

      val key: PurchaseKey = new PurchaseKey(id, new UserKey(user.login))

      val value: Purchase = new Purchase(user.login, game, isTwoPlayer)

      if (i % 10 == 0) logger info s"Sending the ${i}th Producer Record"

      val record = new ProducerRecord[PurchaseKey, Purchase]("avro-user-purchases", key, value)

      purchaseProducer.send(record)
    }

    i += 1
  }

}
